File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/DeepLinkStaticTabDialog.cs
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Connector;
using Microsoft.Teams.TemplateBotCSharp.Properties;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Threading.Tasks;

namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
{
    /// <summary>
    /// This is DeepLink Dialog Class. Main purpose of this class is to show Deep link from Bot to Tab example
    /// </summary>

    [Serializable]
    public class DeepLinkStaticTabDialog : IDialog<object>
    {
        private const string TabEntityID = "statictab";
        private const string TabConfigEntityID = "configTab";
        private string BotId { get; set; }
        private bool IsChannelUser { get; set; } = false;
        private string ChannelId { get; set; }
        private string TabUrl { get; set; }
        private string ButtonCaption { get; set; }
        private string DeepLinkCardTitle { get; set; }

        public async Task StartAsync(IDialogContext context)
        {
            if (context == null)
            {
                throw new ArgumentNullException(nameof(context));
            }

            BotId = ConfigurationManager.AppSettings["MicrosoftAppId"];

            GetChannelID(context);

            var message = CreateDeepLinkMessage(context);

            //Set the Last Dialog in Conversation Data
            context.UserData.SetValue(Strings.LastDialogKey, Strings.LastDialogDeepLinkStaticTabDialog);

            await context.PostAsync(message);

            context.Done<object>(null);
        }

        #region Create Deep Link Tab Card
        private IMessageActivity CreateDeepLinkMessage(IDialogContext context)
        {
            var message = context.MakeMessage();
            var attachment = CreateDeepLinkCard();
            message.Attachments.Add(attachment);
            return message;
        }

        private Attachment CreateDeepLinkCard()
        {
            if (IsChannelUser)
            {
                TabUrl = GetConfigTabDeepLinkURL(ChannelId);
                ButtonCaption = Strings.DeepLinkCardConfigButtonCaption;
                DeepLinkCardTitle = Strings.DeepLinkCardConfigTitle;
            }
            else
            {
                TabUrl = GetStaticTabDeepLinkURL();
                ButtonCaption = Strings.DeepLinkCard1To1ButtonCaption;
                DeepLinkCardTitle = Strings.DeepLinkCard1To1Title;
            }

            return new HeroCard
            {
                Title = DeepLinkCardTitle,
                Buttons = new List<CardAction>
                {
                   new CardAction(ActionTypes.OpenUrl, ButtonCaption, value: TabUrl),
                }
            }.ToAttachment();
        }

        private string GetStaticTabDeepLinkURL()
        {
            //Example -  BaseURL + 28:BotId + TabEntityId (set in the manifest) + ?conversationType=chat
            return "https://teams.microsoft.com/l/entity/28:" + BotId + "/" + TabEntityID + "?conversationType=chat";
        }

        private string GetConfigTabDeepLinkURL(string channelId)
        {
            //Example -  BaseURL + BotId + TabConfigEntityId (e.g. entityId: "configTab" : it should be same which we have set at the time of Tab Creation like below) + ?context= + {"channelId":"19:47051e5643ed49b58665e1250b6db460@thread.skype"} (should be encoded)
            //microsoftTeams.settings.setSettings({ suggestedDisplayName: "Bot Info", contentUrl: createTabUrl(), entityId: "configTab" });

            channelId = channelId.Replace("19:", "19%3a")
                                     .Replace("@thread.skype", "%40thread.skype");

            return "https://teams.microsoft.com/l/entity/" + BotId + "/" + TabConfigEntityID + "?context=%7B%22channelId%22%3A%22" + channelId + "%22%7D";
        }

        private void GetChannelID(IDialogContext context)
        {

            IsChannelUser = false;

            if (context.Activity.ChannelData != null)
            {
                ChannelId = context.Activity.ChannelData["teamsChannelId"];

                if (!string.IsNullOrEmpty(ChannelId))
                {
                    IsChannelUser = true;
                }
                else
                {
                    IsChannelUser = false;
                }
            }
        }
        #endregion
    }
}

File: samples/app-complete-sample/nodejs/server/dialogs/teams/deepLinkStaticTabDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { CardFactory, ActionTypes } = require('botbuilder');
const { WaterfallDialog, ComponentDialog } = require('botbuilder-dialogs');
const DEEPLINKTAB = 'DeepLinkTab';
const TabEntityID = "statictab";
const TabConfigEntityID = "configTab";
let isChannelUser;
let channelId;
let tabUrl;
let buttonCaption;
let deepLinkCardTitle;
let botId;

/**
 * DeepLinkStaticTabDialog class extends ComponentDialog to handle deep link to static tab interactions.
 */
class DeepLinkStaticTabDialog extends ComponentDialog {
    /**
     * Constructor for the DeepLinkStaticTabDialog class.
     * @param {string} id - The dialog ID.
     * @param {StatePropertyAccessor} conversationDataAccessor - The state property accessor for conversation data.
     */
    constructor(id, conversationDataAccessor) {
        super(id);
        this.conversationDataAccessor = conversationDataAccessor;
        // Define the conversation flow using a waterfall model.
        this.addDialog(new WaterfallDialog(DEEPLINKTAB, [
            this.beginDeepLinkStaticTabDialog.bind(this),
        ]));
    }

    /**
     * Begins the deep link static tab dialog.
     * @param {WaterfallStepContext} stepContext - The waterfall step context.
     * @returns {Promise<DialogTurnResult>} The result of the dialog turn.
     */
    async beginDeepLinkStaticTabDialog(stepContext) {
        const currentState = await this.conversationDataAccessor.get(stepContext.context, {});
        currentState.lastDialogKey = "DeeplinkDialog";
        botId = process.env.MicrosoftAppId;
        this.getChannelID(stepContext);
        const message = this.createDeepLinkMessage(stepContext);
        await stepContext.context.sendActivity(message);
        return await stepContext.endDialog();
    }

    /**
     * Creates and returns a deep link message.
     * @param {WaterfallStepContext} stepContext - The waterfall step context.
     * @returns {Partial<Activity>} The deep link message.
     */
    createDeepLinkMessage(stepContext) {
        const reply = stepContext.context._activity;

        if (reply.attachments != null && reply.entities.length > 1) {
            reply.attachments = null;
            reply.entities.splice(0, 1);
        }

        const card = this.createDeepLinkCard();
        reply.attachments = [card];
        return reply;
    }

    /**
     * Creates and returns a deep link card.
     * @returns {Attachment} The deep link card attachment.
     */
    createDeepLinkCard() {
        if (isChannelUser) {
            tabUrl = this.getConfigTabDeepLinkURL(channelId);
            buttonCaption = "Config Tab Deep Link";
            deepLinkCardTitle = "Please click below to navigate config tab";
        } else {
            tabUrl = this.getStaticTabDeepLinkURL();
            buttonCaption = "Static Tab Deep Link";
            deepLinkCardTitle = "Please click below to navigate static tab";
        }

        const buttons = [
            { type: ActionTypes.OpenUrl, title: buttonCaption, value: tabUrl }
        ];

        return CardFactory.heroCard(deepLinkCardTitle, undefined, buttons);
    }

    /**
     * Returns the static tab deep link URL.
     * @returns {string} The static tab deep link URL.
     */
    getStaticTabDeepLinkURL() {
        return `https://teams.microsoft.com/l/entity/28:${botId}/${TabEntityID}?conversationType=chat`;
    }

    /**
     * Returns the config tab deep link URL.
     * @param {string} channelId - The channel ID.
     * @returns {string} The config tab deep link URL.
     */
    getConfigTabDeepLinkURL(channelId) {
        channelId = channelId.replace("19:", "19%3a").replace("@thread.skype", "%40thread.skype");
        return `https://teams.microsoft.com/l/entity/${botId}/${TabConfigEntityID}?context=%7B%22channelId%22%3A%22${channelId}%22%7D`;
    }

    /**
     * Sets the channel ID and determines if the user is in a channel.
     * @param {WaterfallStepContext} stepContext - The waterfall step context.
     */
    getChannelID(stepContext) {
        isChannelUser = false;

        if (stepContext.context._activity.channelData != null) {
            channelId = stepContext.context._activity.channelData.teamsChannelId;

            if (channelId != null) {
                isChannelUser = true;
            }
        }
    }
}

exports.DeepLinkStaticTabDialog = DeepLinkStaticTabDialog;